### PR TITLE
added create2transfer

### PR DESCRIPTION
**Pull request by vaibhavchellani**

### CODE REVIEW

File: contracts/libs/Types.sol [hunk 1]
```diff
@@ -157,6 +156,17 @@ library Types {
...
+        uint256 txType;
+        uint256 fromIndex;
+        uint256 toIndex;
+        uint256 toAccID;
+        uint256 tokenType;
+        uint256 nonce;
+        uint256 amount;
+        uint256 fee;
+    }
+
```
Comment (vaibhavchellani) [inline contracts/libs/Types.sol:169]: The only difference is the addition of `toAccID`

File: contracts/Create2Transfer.sol [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        // Validate we are creating on a zero account
+        require(
+            MerkleTreeUtilsLib.verifyLeaf(
+                stateRoot,
+                RollupUtilsLib.HashFromAccount(toAccountProof.account),
+                _tx.toIndex,
+                toAccountProof.siblings
+            ),
+            "Transfer: receiver proof invalid"
+        );
```
Comment (vaibhavchellani) [inline contracts/Create2Transfer.sol:215]: We allow anyone to prove the previous state of the leaf

File: contracts/Create2Transfer.sol [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        (new_from_account, newRoot) = ApplyTransferTxSender(
+            fromAccountProof,
+            _tx
+        );
+
+        // Validate we are creating on a zero account
+        require(
+            MerkleTreeUtilsLib.verifyLeaf(
+                stateRoot,
+                keccak256(abi.encode(0)),
```
Comment (vaibhavchellani) [inline contracts/Create2Transfer.sol:210]: Prove account is 0 before state transition application on receiver.

File: contracts/Create2Transfer.sol [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+    function ApplyTransferTxReceiver(
+        Types.AccountMerkleProof memory _merkle_proof,
+        Tx.Create2Transfer memory _tx,
+        uint256 token
+    ) public pure returns (bytes memory updatedAccount, bytes32 newRoot) {
+        // Initialize account
+        _merkle_proof.account.ID = _tx.toAccID;
+        _merkle_proof.account.balance = _tx.amount;
+        _merkle_proof.account.tokenType = token;
```
Comment (vaibhavchellani) [inline contracts/Create2Transfer.sol:256]: Initialize receiver account

File: contracts/libs/RollupUtils.sol [hunk 5]
```diff
@@ -173,4 +168,119 @@ contract RollupUtils {
...
+            transaction.txType,
+            ,
+            ,
+            toAccID,
+            transaction.nonce,
+            transaction.amount,
+            transaction.fee
+        ) = abi.decode(
+            txBytes,
+            (uint256, uint256, uint256, uint256, uint256, uint256, uint256)
```
Comment (ChihChengLiang) [inline contracts/libs/RollupUtils.sol:258]: pubkey should be uint256[4]?

File: contracts/Create2Transfer.sol [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+                MerkleTreeUtilsLib.verifyLeaf(
+                    accountRoot,
+                    keccak256(abi.encodePacked(proof.pubkeysSender[i])),
+                    proof.states[i].pubkeyIndex,
+                    proof.pubkeyWitnessesSender[i]
+                ),
+                "Rollup: from account does not exists"
+            );
+
+            // check pubkey inclusion
```
Comment (ChihChengLiang) [inline contracts/Create2Transfer.sol:51]: check receiver pubkey inclusion

File: contracts/Create2Transfer.sol [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        (new_to_account, newRoot) = ApplyTransferTxReceiver(
+            to,
+            _tx,
+            from.state.tokenType
+        );
+
+        return (newRoot, new_from_account, new_to_account, Types.Result.Ok);
+    }
+
+    function ApplyTransferTxSender(
```
Comment (ChihChengLiang) [inline contracts/Create2Transfer.sol:198]: Maybe a rename?

File: contracts/Create2Transfer.sol [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    ) public pure returns (bytes memory updatedAccount, bytes32 newRoot) {
+        // Initialize account
+        Types.UserState memory newState = Types.UserState(
+            _tx.toAccID,
+            token,
+            _tx.amount,
+            0
+        );
+
+        bytes memory accountInBytes = newState.encode();
```
Comment (ChihChengLiang) [inline contracts/Create2Transfer.sol:227]: encodedState

File: test/create2transfer.test.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+            pubkeyWitnessesSender.push(registry.witness(sender.pubkeyIndex));
+
+            pubkeysReceiver.push(receiver.getPubkey());
+            pubkeyWitnessesReceiver.push(
+                registry.witness(receiver.pubkeyIndex)
+            );
+
+            const signedObj = sender.signAndReturnMessage(tx);
+            signatures.push(signedObj.signature);
+            message.push(signedObj.M);
```
Comment (ChihChengLiang) [inline test/create2transfer.test.ts:101]: message is unused

File: test/create2transfer.test.ts [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        const signature = mcl.aggreagate(signatures);
+        let res = await bls.verifyMultiple(signature, pubkeysSender, message);
```
Comment (ChihChengLiang) [inline test/create2transfer.test.ts:105]: can remove this too

File: test/rollupUtils.test.ts [hunk 11]
```diff
@@ -58,4 +61,35 @@ describe("RollupUtils", async function() {
...
+        ](encodedTx);
+
+        assert.equal(
+            decodedTx.fromIndex.toString(),
+            "3",
+            "from index not correct"
+        );
+        assert.equal(decodedTx.toIndex.toString(), "4", "to index not correct");
+        // TODO fix
+        // assert.equal(decodedTx.amount.toString(), "1", "amount not correct");
```
Comment (ChihChengLiang) [inline test/rollupUtils.test.ts:93]: maybe try this again?

File: ts/tx.ts [hunk 12]
```diff
@@ -66,23 +73,39 @@ export class TxTransfer implements SignableTx {
...
+                this.fromIndex,
+                this.toIndex,
+                this.nonce,
+                this.amount.toString(),
+                this.fee
+            ]
+        );
+    }
+
+    public hash(): string {
```
Comment (ChihChengLiang) [inline ts/tx.ts:89]: we no longer have hash()

File: ts/tx.ts [hunk 13]
```diff
@@ -169,3 +191,107 @@ export class TxMassMigration implements SignableTx {
...
+                this.fromPubkey,
+                this.toPubkey,
+                this.nonce,
+                this.amount.toString(),
+                this.fee
+            ]
+        );
+    }
+
+    public hash(): string {
```
Comment (ChihChengLiang) [inline ts/tx.ts:262]: no hash()

File: ts/state.ts [hunk 14]
```diff
@@ -62,15 +64,25 @@ export class State {
 
     public sign(tx: SignableTx) {
         const msg = tx.message();
-        const { signature, M } = mcl.sign(msg, this.secretKey);
+        const { signature } = mcl.sign(msg, this.secretKey);
         return signature;
     }
 
+    public signAndReturnMessage(tx: SignableTx) {
```
Comment (ChihChengLiang) [inline ts/state.ts:71]: we can remove this now

File: ts/stateTree.ts [hunk 15]
```diff
@@ -123,6 +142,31 @@ export class StateTree {
...
+        for (let i = 0; i < txs.length; i++) {
+            if (safe) {
+                const proof = this.applyTxCreate2Transfer(txs[i]);
+                proofs.push(proof);
+                safe = proof.safe;
+            } else {
+                proofs.push(PLACEHOLDER_TRANSFER_PROOF);
+            }
+        }
+        const sumOfFee = txs.map(tx => tx.fee).reduce((a, b) => a.add(b));
```
Comment (ChihChengLiang) [inline ts/stateTree.ts:164]: can use sum() from the utils

File: ts/stateTree.ts [hunk 16]
```diff
@@ -246,4 +290,55 @@ export class StateTree {
...
+        // create receiver account
+        const receiverState = State.new(
+            tx.toAccID,
+            senderState.tokenType,
+            0,
+            0
+        );
+
+        receiverState.balance = receiverState.balance.add(tx.amount);
+        receiverState.stateID = tx.toAccID;
```
Comment (ChihChengLiang) [inline ts/stateTree.ts:330]: tx.toIndex?

File: ts/stateTree.ts [hunk 17]
```diff
@@ -246,4 +290,55 @@ export class StateTree {
...
+        const receiverStateStruct = receiverState.toSolStruct();
+        // TODO add pubkey from account tree
```
Comment (ChihChengLiang) [inline ts/stateTree.ts:332]: can remove

File: ts/state.ts [hunk 18]
```diff
@@ -36,6 +36,8 @@ export class State {
             nonce
         );
     }
+
+    // TODO add optional params for pubkey and stateID
```
Comment (ChihChengLiang) [inline ts/state.ts:40]: let's create an issue for this instead of checking in a TODO comment

**Review by ChihChengLiang — APPROVED** [review]
Great work! Looks good to me

**Review by ChihChengLiang — APPROVED** [review]
Epic work! Excited to see it merged.

**Pull request merged**